impedance/validation.py

```diff
@@ -77,85 +88,197 @@ def linKK(f, Z, c=0.85, max_M=50):
-                           args=(R0, ts, Z, f, 'both'),
-                           ftol=1E-13, gtol=1E-10)
+    w = 2 * np.pi * f
+
+    # Fitting model has M RC elements plus 1 series resistance and 1 series
+    # inductance
+    a_re = np.zeros((f.size, M+2))
+    a_im = np.zeros((f.size, M+2))
+
+    if add_cap:
+        a_re = np.zeros((f.size, M+3))
+        a_im = np.zeros((f.size, M+3))
+
+        # Column for series capacitance. Real part = 0.
+        a_im[:, -2] = - 1 / (w * np.abs(Z))
+
+    # Column for series resistance, R_o in model. Imaginary part = 0.
+    a_re[:, 0] = 1 / np.abs(Z)
+
+    # Column for series inductance to capture inevitable contributions from
+    # the measurement system. Real part = 0.
+    a_im[:, -1] = w / np.abs(Z)
+
+    # Columns for series RC elements
+    for i, tau in enumerate(ts):
+        a_re[:, i+1] = K([1, tau], f).real / np.abs(Z)
+        a_im[:, i+1] = K([1, tau], f).imag / np.abs(Z)
+
+    if fit_type == 'real':
+        p_values = np.linalg.pinv(a_re).dot(Z.real / np.abs(Z))
+
+        # After fitting real part, need to use imaginary component of fit to
+        # find values of series inductance and capacitance
+        a_im = np.zeros((f.size, 2))
+        a_im[:, -1] = w / np.abs(Z)
+        if add_cap:
+            a_im[:, -2] = -1 / (w * np.abs(Z))
+            p_values[-2] = 1e-18  # nullifies series C without dividing by 0
+
+        Z_fit_re = eval_linKK(p_values, ts, f)
+        coefs = np.linalg.pinv(a_im).dot((Z.imag - Z_fit_re.imag) / np. abs(Z))
+
+        if add_cap:
+            p_values[-2:] = coefs
+        else:
+            p_values[-1] = coefs[-1]
+    elif fit_type == 'imag':
+        p_values = np.linalg.pinv(a_im).dot(Z.imag / np.abs(Z))
+
+        # Calculates real part of impedance from fitting to imaginary parts
+        # without ohmic resistance, i.e. only the real parts of series RC
+        # elements.
+        z_re = eval_linKK(p_values, ts, f)
+
+        # Weighting used in Boukamp et al - "A Linear Kronig-Kramers
+        # Transform for Immittance Data Validation" 1995, J. Electrochem
+        # Soc. 142 (6)
+        ws = 1 / (Z.real**2 + Z.imag**2)
+
+        # Finds ohmic resistance for imaginary part fit according to Eq 7
+        # of Boukamp et al.
+        p_values[0] = np.sum(ws * (Z.real - z_re.real)) / np.sum(ws)
+    elif fit_type == 'complex':
+        x = np.linalg.inv(a_re.T.dot(a_re) + a_im.T.dot(a_im))
+        y = a_re.T.dot(Z.real / np.abs(Z)) + a_im.T.dot(Z.imag / np.abs(Z))
+        p_values = x.dot(y)
+    else:
+        raise ValueError("Invalid choice of fit_type, please choose from "
+                         "\'real\', \'imag\', or \'complex\'")
 
-    p_values = result['x']
-    mu = calc_mu(p_values)
+    if add_cap:
+        mu = calc_mu(p_values[:-2])
+    else:
+        mu = calc_mu(p_values[:-1])
 
     return p_values, mu
 
 
-def eval_linKK(Rs, R0, ts, f):
+def eval_linKK(Rs, ts, f):
```

Comment:
Smallest of suggestions here, but maybe makes sense to rename Rs -> elements (or something like that?) It was a little confusing sec for me to see Rs[-1] get substituted into L and C 😝. 

We've also now bumped the minimum supported Python up to 3.7 so these could be f-strings if that's easier (e.g. `f'L({Rs[-1]},{f.tolist()})'`.

Reply:
Yeah, I agree. Maybe changing any reference of `Rs` to `p_values` so it's consistent with outputs of `fit_linKK`? I guess one exception `calculate_mu` where only the resistances should be passed. I also just realized R_ohmic is currently used in calculating mu and that's probably skewing it to higher values. I'll fix that in the same commit.